segmentation/utils/synthetic_spatial_datagen.py

```diff
@@ -0,0 +1,204 @@
+
+    Each row and column representing a cell.
+    We generate the points using Gaussian distributions
+    Ideally, the parameters for A to B distances will be set such that they produce a lower range of values
+    than A to C distances.
+
+    Will return a random distance matrix as well as the dictionary of associated cell: label IDs
+    The above is generated from the generate_labels function
+
+    Args:
+        num_A: the number of A cells we wish to generate. Default 100
+        num_B: the number of B cells we wish to generate. Default 100
+        num_C: the number of C cells we wish to generate. Default 100
+        distr_AB: if specified, will be a dict listing the mean and variance of the Gaussian distribution
+            we wish to generate numbers from. If None, use the default values.
+        distr_AC: similar to dist_AB. Default will have a higher mean value.
+        seed: whether to fix the random seed or not. Useful for testing.
+            Should be a specified integer value. Default None.
+    """
+
+    # set the mean and variance of the Gaussian distributions of both AB and AC distances
+    if not distr_AB:
+        mean_ab = AB_DIST_MEAN
+        var_ab = AB_DIST_VAR
+    else:
+        mean_ab = distr_AB['mean']
+        var_ab = distr_AB['var']
+
+    if not distr_AC:
+        mean_ac = AC_DIST_MEAN
+        var_ac = AC_DIST_VAR
+    else:
+        mean_ac = distr_AC['mean']
+        var_ac = distr_AC['var']
+
+    # set random seed if set
+    if seed:
+        np.random.seed(seed)
+
+    # we initialize the random distances across different types of points
+    # note that we don't really care about aa, bb, bc, or cc, so we
+    # initialize those to garbage. We do need them for a proper
+    # distance matrix format, however.
+    random_aa = np.random.normal(0, 1, (num_A, num_A))
+    random_ab = np.random.normal(mean_ab, var_ab, (num_A, num_B))
+    random_ac = np.random.normal(mean_ac, var_ac, (num_A, num_C))
+    random_bb = np.random.normal(0, 1, (num_B, num_B))
+    random_bc = np.random.normal(0, 1, (num_B, num_C))
+    random_cc = np.random.normal(0, 1, (num_C, num_C))
+
+    # create each row one-by-one first
+    # we need to correct each aa, bb, and cc matrix to ensure symmetry
+    first_row = np.concatenate(((random_aa + random_aa.T) / 2, random_ab, random_ac), axis=1)
+    second_row = np.concatenate((random_ab.T, (random_bb + random_bb.T) / 2, random_bc), axis=1)
+    third_row = np.concatenate((random_ac.T, random_bc.T, (random_cc + random_cc.T) / 2), axis=1)
+
+    # then concatenate them together
+    dist_mat = np.concatenate((first_row, second_row, third_row), axis=0)
+
+    # assert that the created submatrix is symmetric
+    assert np.allclose(dist_mat, dist_mat.T, rtol=1e-05, atol=1e-08)
+
+    return dist_mat
+
+
+def get_random_centroid_centers(size_img=(1024, 1024), num_A=100, num_B=100, num_C=100, distr_A=None, distr_B=None, distr_C=None, seed=None):
```

Comment:
instead of inputting a matrix for the covariance, you could use a tuple (sigx, sigy, theta) and generate the covariance matrix with something like:
```
# generate scaling
scale = np.array([[sigx**2, 0], [0, sigy**2]])

# generate rotation matrix
s, c = np.sin(theta), np.cos(theta)
rot = np.array([[c, -s], [s, c]])

# compute covariance matrix
cov = rot * scale * rot.T
```
this guarantees that the final matrix is symmetric positive definite, and covers all possible cases; could maybe go into a separate function too